File: tutorial1/NebulaTutorial1/TodoListViewController.m
//
//  ViewController.m
//  NebulaTutorial1
//  Copyright (c) 2014-2018, NEC Corporation. All rights reserved.
//

#import "nebulaIosSdk/Nebula.h"

#import "TodoListViewController.h"
#import "TodoEditViewController.h"

#define BUCKET_NAME @"TodoTutorial1"

@interface TodoListViewController ()
{
    UIActivityIndicatorView *indicatorView;
}
@property(nonatomic) NSArray *todos;
@end

@implementation TodoListViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [self.tableView registerClass:[UITableViewCell class] forCellReuseIdentifier:@"TodoCell"];
    indicatorView = nil;
    
    self.todos = @[]; // initial
    [self reloadTodos:nil];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
}

// Todo 一覧を取得する
- (IBAction)reloadTodos:(id)sender {
    NBObjectBucket *bucket = [[NBObjectBucket alloc] initWithBucketName:BUCKET_NAME];

    NBQuery *query = [NBQuery new];
    [query setSortOrderWithKey:@"updatedAt" isAscend:YES];
    
    [bucket queryInBackgroundWithQuery:query block:^(NSArray *objects, NSUInteger count, NSError *error) {
        [self hideIndicator];
        if (!error) {
            self.todos = objects;
            [self.tableView reloadData];
        } else {
            NSLog(@"query error: %@", error);
        }
    }];
    [self showIndicator];
}

#pragma mark - UITableViewDataSource

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return (self.todos).count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    UITableViewCell *cell = [self.tableView dequeueReusableCellWithIdentifier:@"TodoCell"];

    NBObject *todo = (self.todos)[indexPath.row];

    cell.textLabel.text = todo[@"description"];
    return cell;
}

- (BOOL)tableView:(UITableView *)tableView canEditRowAtIndexPath:(NSIndexPath *)indexPath
{
    return YES; // 削除可能
}

/**
 * 削除処理
 */
- (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath
{
    NBObject *obj = (self.todos)[indexPath.row];

    // オブジェクト削除
    [obj deleteInBackgroundWithBlock:^(NSError *error) {
        [self hideIndicator];
        if (!error) {
            [self reloadTodos:nil];
        } else {
            NSLog(@"Delete failed: %@", error);
        }
    }];
    [self showIndicator];
}

#pragma mark - UITableViewDelegate

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [tableView deselectRowAtIndexPath:indexPath animated:NO];
}

#pragma mark - Add Todo

- (IBAction)firstViewReturnActionForSegue:(UIStoryboardSegue *)segue
{
    if ([segue.identifier isEqualToString:@"done"]) {
        TodoEditViewController *vc = segue.sourceViewController;
        NSString *todo = vc.todoText;
        [self addTodo:todo];
    }
}

/**
 * Todo の追加処理
 */
- (void)addTodo:(NSString *)description {
    // オブジェクト作成
    NBObject *todo = [NBObject objectWithBucketName:BUCKET_NAME];

    // オブジェクトに値を設定
    todo[@"description"] = description;

    // オブジェクト保存
    [self showIndicator];
    [todo saveInBackgroundWithBlock:^(NSArray *objects, NSUInteger count, NSError *error) {
        [self hideIndicator];
        if (!error) {
            [self reloadTodos:nil];
        } else {
            NSLog(@"save error : %@", error);
        }
    }];
}

- (void)showIndicator
{
    if (indicatorView == nil) {
        indicatorView = [[UIActivityIndicatorView alloc] initWithFrame:CGRectMake(0, 0, self.view.bounds.size.width, self.view.bounds.size.height)];
        indicatorView.activityIndicatorViewStyle = UIActivityIndicatorViewStyleWhiteLarge;
        indicatorView.backgroundColor = [UIColor colorWithWhite:0.0 alpha:0.3];
        indicatorView.contentMode = UIViewContentModeCenter;
        indicatorView.autoresizingMask =
            UIViewAutoresizingFlexibleLeftMargin |
            UIViewAutoresizingFlexibleRightMargin |
            UIViewAutoresizingFlexibleTopMargin |
            UIViewAutoresizingFlexibleBottomMargin;
        [self.view addSubview:indicatorView];
        [indicatorView startAnimating];
    }
}

- (void)hideIndicator
{
    if (indicatorView != nil) {
        [indicatorView stopAnimating];
        [indicatorView removeFromSuperview];
        indicatorView = nil;
    }
}

@end

File: tutorial1-swift/NebulaTutorial1/TodoListViewController.swift
//
//  TodoListViewController.swift
//  Copyright (c) 2014年 NEC Corporation. All rights reserved.
//

import UIKit

class TodoListViewController: UITableViewController {
    
    let BUCKET_NAME = "TodoTutorial1";
    
    var indicatorView: UIActivityIndicatorView? = nil
    
    var todos: [NBObject] = []

    override func viewDidLoad() {
        super.viewDidLoad()
        tableView.register(UITableViewCell.self, forCellReuseIdentifier: "TodoCell")
        
        reloadTodos(nil)
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
    }
    
    // Todo 一覧を取得する
    @IBAction func reloadTodos(_ sender: UIButton?) {

        // バケット取得
        let bucket = NBObjectBucket(bucketName: BUCKET_NAME)
        
        // クエリ生成
        let query = NBQuery()
        query.setSortOrder(key: "updatedAt", isAscend: true)

        // クエリ実行
        bucket.queryInBackground(query: query) { objects, count, error in
            self.hideIndicator()
            guard error == nil else {
                print("query error: \(error!)")
                return
            }
            self.todos = objects!
            self.tableView.reloadData()
        }
        showIndicator()
    }
    
    override func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }
    
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return todos.count;
    }

    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "TodoCell")!
        
        let todo = todos[indexPath.row]
        cell.textLabel?.text = todo.object(forKey: "description") as? String
        
        return cell
    }
    
    override func tableView(_ tableView: UITableView, canEditRowAt indexPath: IndexPath) -> Bool {
        return true // 削除可能
    }
    
    /**
    * 削除処理
    */
    override func tableView(_ tableView: UITableView, commit editingStyle: UITableViewCellEditingStyle, forRowAt indexPath: IndexPath) {
        let obj = todos[indexPath.row]

    
        // オブジェクト削除
        obj.deleteInBackground() { error in
            self.hideIndicator()
            guard error == nil else {
                print("Delete failed: \(error!)")
                return
            }
            self.reloadTodos(nil)
        }
        showIndicator()
    }
    
    // UITableViewDelegate
    
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: false)
    }
    
    // pragma mark - Add Todo
    
    @IBAction func firstViewReturnActionForSegue(_ segue: UIStoryboardSegue) {
        if (segue.identifier == "done") {
            let vc = segue.source as! TodoEditViewController
            let todo = vc.todoText
            addTodo(todo)
        }
    }
    
    /**
    * Todo の追加処理
    */
    func addTodo(_ description: String) {
        // オブジェクト作成
        let todo = NBObject(bucketName: BUCKET_NAME)
    
        // オブジェクトに値を設定
        todo.setObject(description, forKey: "description")
    
        // オブジェクト保存
        showIndicator()
        
        todo.saveInBackground() { objects, count, error in
            self.hideIndicator()
            guard error == nil else {
                print("save error : \(error!)")
                return
            }
            self.reloadTodos(nil)
        }
    }

    /**
     * IndicatorView を表示する
     */
    func showIndicator() {
        if (indicatorView != nil) {
            return
        }
        
        let v = UIActivityIndicatorView(frame:
            CGRect(x:0, y:0, width:self.view.bounds.size.width, height:self.view.bounds.size.height))
        
        v.activityIndicatorViewStyle = UIActivityIndicatorViewStyle.whiteLarge;
        v.backgroundColor = UIColor(white:0.0, alpha:0.3)
        v.contentMode = UIViewContentMode.center;
        v.autoresizingMask =
                [.flexibleLeftMargin, .flexibleRightMargin, .flexibleTopMargin, .flexibleBottomMargin]

        self.view.addSubview(v)
        v.startAnimating()
        
        self.indicatorView = v
    }
    
    /**
     * IndicatorView を消す
     */
    func hideIndicator() {
        if (indicatorView == nil) {
            return
        }

        indicatorView!.stopAnimating()
        indicatorView!.removeFromSuperview()
        indicatorView = nil;
    }
}

